src/libstd/sync/mpsc/sync.rs

```diff
@@ -126,6 +127,28 @@ fn wait<'a, 'b, T>(lock: &'a Mutex<State<T>>,
     lock.lock().unwrap() // relock
 }
 
+/// Same as wait, but waiting at most `timeout`.
+fn wait_timeout_receiver<'a, 'b, T>(lock: &'a Mutex<State<T>>,
+                                    timeout: Duration,
+                                    mut guard: MutexGuard<'b, State<T>>)
+                                    -> MutexGuard<'a, State<T>>
+{
+    let (wait_token, signal_token) = blocking::tokens();
+    match mem::replace(&mut guard.blocker, BlockedReceiver(signal_token)) {
+        NoneBlocked => {}
+        _ => unreachable!(),
+    }
+    drop(guard);         // unlock
+    let timed_out = !wait_token.wait_timeout(timeout);   // block
```

Comment:
This probably wants to ignore the return value of `wait_timeout` and instead just test the internal state to resolve any races between timing out and sending a message near the timeout.